### DISCUSSION TITLE
Try to use WIFI MAC address as deveui

### DISCUSSION THREAD

**Discussion by daeynasvistas** [ngraziano/LMICPP-Arduino#24] [General]
ok.. so in other lib i use the MAC address from ESP as "deveui" and "ssid"

something like this:

```
char * fooMAC()
{
  char ssid_mac [64];
  uint8_t mac[6];
  WiFi.macAddress(mac);
  snprintf( ssid_mac, sizeof( ssid_mac ), "%02x%02x%02x%02x%02x%02x%02x%02x", 02,00, mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
  return ssid_mac;
} 

const char * ssid = fooMAC();
```
but now i need to pass char const *DEVEUI:
```
SetupLmicKey<appEui, deveui, appKey>::setup(LMIC);
```

i try to do this, with no luck..
```
char deveui_mac [64];


void fooMAC()
{
  uint8_t mac[6];
  WiFi.macAddress(mac);
  snprintf( deveui_mac, sizeof( deveui_mac ), "%02x%02x%02x%02x%02x%02x%02x%02x", 02,00, mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
  //return deveui_mac;
} 
(...)
  SetupLmicKey<appEui, deveui_mac, appKey>::setup(LMIC);
```

deveui are always 
devEUI:"0000000000000000"

i know its nothing related with lib.. but if someone can help..

**Comment by ngraziano** [comment]
To use SetupLmicKey the key must be known at compile time. But you can always call LMIC.setDevEuiCallback  after to replace with the callback you want (before starting to send data). The callback must fill buffer low byte first

Something like that (I have not check if it work)
```cpp

void getEui(uint8_t *buf) { 
// I am not sure if it is in the good order
  buf[0] = xx;
  buf[1]= yy;
  WiFi.macAddress(buf+2);

}



void setup() {
(...)
SetupLmicKey<appEui, fakedeveui, appKey>::setup(LMIC);
LMIC.setDevEuiCallback(getEui);
(...)
}
```

**Comment by daeynasvistas** [comment]
its great...

```
void getEui(uint8_t *buf) { 
  uint8_t mac[6];  
  WiFi.macAddress(mac);

  buf[0] = mac[5],HEX;
  buf[1] = mac[4],HEX;
  buf[2] = mac[3],HEX;
  buf[3] = mac[2],HEX;
  buf[4] = mac[1],HEX;
  buf[5] = mac[0],HEX;
  buf[6] = 0,HEX;
  buf[7] = 2,HEX;
}
```

So SSID from WIFI AP are the same that LoRa Deveui 
in "HEX format".

ESP Board MAC Address:  (02:00)4C:75:25:C1:C8:44
![image](https://user-images.githubusercontent.com/2634610/155143302-8466d98f-0743-495f-8ef3-e16f5f398fc2.png)
